### PR TITLE
Drewsetski component notification

### PR DESCRIPTION
**Pull request by hakankaradis**
Reopen: #9757

## Screenshots (if appropriate):
http://recordit.co/csI5NjqBLC

## Types of changes
- [x] New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: client/component-lab/Notification/helpers.coffee [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    clearTimeout timerId
+
+
+  @resume()
+
+validateProps = (notification, types) ->
+
+  if not notification.type
+    throw new Error('Notification type is required.')
+  if _.keys(types).indexOf(notification.type) is -1
```
Comment (usirin) [inline client/component-lab/Notification/helpers.coffee]: Use `Object.keys`

File: client/component-lab/Notification/helpers.coffee [hunk 2]
```diff
@@ -0,0 +1,38 @@
+_ = require 'lodash'
+
+Timer = (callback, delay) ->
```
Comment (usirin) [inline client/component-lab/Notification/helpers.coffee]: Make this a proper `class`.

File: client/component-lab/Notification/helpers.coffee [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+  timerId = undefined
+  start = undefined
```
Comment (usirin) [inline client/component-lab/Notification/helpers.coffee]: Is there a specific reason why `null` is not being used here?

File: client/component-lab/Notification/NotificationViewContainer.coffee [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    { enter, leave } = @getAnimationProps()
+
+    stateClass = if @options.notifications.length
+    then ''
+    else 'hidden'
+
+    className = generateClassName [
+      styles.kd_notification_list
+      stateClass
+    ]
```
Comment (usirin) [inline client/component-lab/Notification/NotificationViewContainer.coffee]: use `classnames` npm module here. ref: https://www.npmjs.com/package/classnames

File: client/component-lab/Notification/NotificationViewContainer.coffee [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+      leave :
+        from :
+          transform : 'scale(1)'
+          opacity : 1
+        to :
+          transform : 'scale(0.9)'
+          opacity : 0
+
+
+  onNotificationRemove: (uid) =>
```
Comment (usirin) [inline client/component-lab/Notification/NotificationViewContainer.coffee]: remove fat arrow.

File: client/component-lab/Notification/NotificationView.coffee [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+      ]
+      iconClass = [
+        styles.kd_notification_icon
+        styles["kd_notification_icon_#{notification.type}"]
+      ]
+      message = notification.content
+      if notification.dismissible
+        className.push styles.kd_notification_dismissible
+      if notification.primaryButtonTitle or notification.secondaryButtonTitle
+        className = className.filter (word) -> word isnt styles.kd_notification_dismissible
```
Comment (usirin) [inline client/component-lab/Notification/NotificationView.coffee]: use `classnames` npm module here. ref: https://www.npmjs.com/package/classnames

File: client/component-lab/Notification/NotificationView.coffee [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+      transitionEvent = getVendorTransition()
+      element = ReactDOM.findDOMNode this
+      notification = @props.notification
+      @mounted = yes
+      if transitionEvent
+        element.addEventListener transitionEvent, @onTransitionEnd
+        if @autoDismissible notification
+          @notificationTimer = new Helpers.Timer =>
+            @hideNotification()
+            return
```
Comment (usirin) [inline client/component-lab/Notification/NotificationView.coffee]: why explicit `return`?

File: client/component-lab/Notification/NotificationView.coffee [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+Helpers = require './helpers'
+Constants = require './constants'
+
+
+module.exports = class NotificationView extends React.Component
+
+    constructor: (props) ->
+
+      super props
+      @style = { zIndex: 100 + @props.index }
```
Comment (usirin) [inline client/component-lab/Notification/NotificationView.coffee]: remove this from here, and define this as `defaultProps.style`

File: client/app/lib/util/getVendorTransition.coffee [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+module.exports = getVendorTransition = ->
+
+  el = document.createElement('fakeelement')
+  transition = null
+  transitions =
+    transition : 'transitionend'
+    OTransition : 'oTransitionEnd'
+    MozTransition : 'transitionend'
+    WebkitTransition : 'webkitTransitionEnd'
+  _.keys(transitions).forEach (transitionKey) ->
```
Comment (usirin) [inline client/app/lib/util/getVendorTransition.coffee]: Use `Object.keys`

**Review by usirin — CHANGES_REQUESTED** [review]
@drewsetski  The requested changes are mostly related to the way code is written. Once they are fixed, it looks mergeable.

**Comment by hakankaradis** [comment]
@sinan and @usirin I have updated this pr with new commits of [this pr](https://github.com/koding/koding/pull/10420)

I have removed https://github.com/koding/koding/pull/10420/files#diff-a51377e3c5c5b024cb7f53f96603be57R28 this style line, it doesn't necessary and also fixed this line of code https://github.com/koding/koding/pull/10420/files#diff-a51377e3c5c5b024cb7f53f96603be57R118

**Comment by sinan** [comment]
@usirin can you please finalize this so that we can start using in the app.

File: client/component-lab/Notification/NotificationViewContainer.coffee (hunk 5 above)
Comment (usirin) [inline client/component-lab/Notification/NotificationViewContainer.coffee]: @hakankaradis this is still not fixed?

**Review by usirin — APPROVED** [review]

**Comment by usirin** [comment]
Minor fat arrow issue, but not a blocker. Thanks @hakankaradis

**Pull request merged**